svgnative/ports/cairo/CairoSVGRenderer.cpp

```diff
@@ -0,0 +1,473 @@
+        break;
+    }
+    return;
+}
+
+inline void appendCairoSvgPath( cairo_t* mCairo, const Path& path )
+{
+    cairo_path_t* cairoPath = getPathObjFromCairoSvgPath(&path);
+    cairo_append_path (mCairo, cairoPath);
+    cairo_path_destroy( cairoPath );
+}
+
+void CairoSVGRenderer::DrawPath(
+    const Path& path, const GraphicStyle& graphicStyle, const FillStyle& fillStyle, const StrokeStyle& strokeStyle)
+{
+    SVG_ASSERT(mCairo);
+    Save(graphicStyle);
+
+
+    double alpha = getAlphaProduct( alphas );
+
+    if (fillStyle.hasFill)
+    {
+        if (fillStyle.paint.type() == typeid(Gradient)) {
+            cairo_pattern_t* pat;
+            createCairoPattern(fillStyle.paint, fillStyle.fillOpacity * alpha, &pat);
+            cairo_set_source(mCairo, pat);
+        } else {
+            const auto& color = boost::get<Color>(fillStyle.paint);
+            cairo_set_source_rgba(mCairo,
+                                  color[0],
+                                  color[1],
+                                  color[2],
+                                  color[3] * fillStyle.fillOpacity * alpha );
+        }
+
+        // Skia backend does not handle fillStyle.Rule yet
+        switch (fillStyle.fillRule) {
+        case WindingRule::kEvenOdd:
+            cairo_set_fill_rule (mCairo, CAIRO_FILL_RULE_EVEN_ODD);
+            break;
+        case WindingRule::kNonZero:
+        default:
+            cairo_set_fill_rule (mCairo, CAIRO_FILL_RULE_WINDING);
+        }
+
+        cairo_new_path (mCairo);
+        appendCairoSvgPath( mCairo, path );
+        cairo_fill (mCairo);
+    }
+    if (strokeStyle.hasStroke)
+    {
+        const auto& color = boost::get<Color>(strokeStyle.paint);
+        cairo_set_source_rgba(mCairo,
+                              color[0],
+                              color[1],
+                              color[2],
+                              color[3] * strokeStyle.strokeOpacity * alpha );
+
+        cairo_set_line_width(mCairo, strokeStyle.lineWidth);
+
+        switch (strokeStyle.lineCap) {
+        case LineCap::kRound:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_ROUND);
+            break;
+        case LineCap::kSquare:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_SQUARE);
+            break;
+        case LineCap::kButt:
+        default:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_BUTT);
+        }
+
+        switch (strokeStyle.lineJoin) {
+	case LineJoin::kRound:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_ROUND);
+            break;
+	case LineJoin::kBevel:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_BEVEL);
+            break;
+	case LineJoin::kMiter:
+	default:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_MITER);
+        }
+
+        // dashArray is not supported in Skia yet (2019/04/28)
+        // copied from String backend
+        double *dashes = NULL;
+        if (!strokeStyle.dashArray.empty()) {
```

Comment:
I'd write it more in CPP style:
```
std::vector<double> dashes;
for (auto dash : strokeStyle.dashArray)
    dash.push_back(dash);
cairo_set_dash(mCairo, dashes.data(), dashes.size(), strokeStyle.dashOffset);
```

Reply:
Thank you very much, I would replace by suggested code.